transaction/core/src/revealed_tx_out.rs

```diff
@@ -0,0 +1,69 @@
+use crate::{tx::TxOut, Amount, AmountError, TxOutConversionError};
+use alloc::{
+    string::{String, ToString},
+    vec::Vec,
+};
+use displaydoc::Display;
+use mc_crypto_digestible::Digestible;
+use prost::Message;
+use serde::{Deserialize, Serialize};
+
+/// A TxOut together with its amount shared secret, which can be used to reveal
+/// the amount and token id and check them against the commitment data
+#[derive(Clone, Digestible, PartialEq, Eq, Message, Serialize, Deserialize)]
+pub struct RevealedTxOut {
+    /// The TxOut which is being revealed
+    #[prost(message, required, tag = "1")]
+    pub tx_out: TxOut,
+
+    /// The amount shared secret of this TxOut
+    /// This should be exactly 32 bytes
+    #[prost(bytes, tag = "2")]
+    pub amount_shared_secret: Vec<u8>,
+}
+
+impl RevealedTxOut {
+    /// Attempt to reveal the amount of this RevealedTxOut
+    pub fn reveal_amount(&self) -> Result<Amount, RevealedTxOutError> {
+        try_reveal_amount(&self.tx_out, self.amount_shared_secret.as_ref())
+    }
+}
+
+/// Helper function which tries to reveal the amount of a TxOut given its shared
+/// secret, and confirm this against the commitment data
+pub fn try_reveal_amount(
+    tx_out: &TxOut,
+    amount_shared_secret: &[u8],
+) -> Result<Amount, RevealedTxOutError> {
+    let ss: &[u8; 32] = amount_shared_secret
+        .try_into()
+        .map_err(|_| RevealedTxOutError::InvalidAmountSharedSecret)?;
+    let (amount, _) = tx_out
+        .get_masked_amount()?
+        .get_value_from_amount_shared_secret(ss)?;
+    Ok(amount)
+}
+
+/// An error that can occur when attempting to reveal a TxOut using its Amount
+/// shared secret
+#[derive(Clone, Debug, Display, Ord, PartialOrd, Hash, Eq, PartialEq, Serialize, Deserialize)]
+pub enum RevealedTxOutError {
+    /// Invalid Amount Shared Secret
+    InvalidAmountSharedSecret,
+    /// TxOut conversion: {0}
+    TxOutConversion(String),
+    /// Amount: {0}
+    Amount(String),
```

Comment:
The problem is that this is used ultimately in `InputRuleError` which becomes part of `TransactionValidationError` which has to convert to `TxProposalResult`, and that has to be convertible back from an error code, so anything that is a member of any of these enums has to be default constructible. See for instance here:

https://github.com/mobilecoinfoundation/mobilecoin/blob/f142750ae329fd94bb0cbc480d4d8019bf182763/consensus/api/src/conversions.rs#L51

https://github.com/mobilecoinfoundation/mobilecoin/blob/f142750ae329fd94bb0cbc480d4d8019bf182763/consensus/api/src/conversions.rs#L107

I think that is the reason that "ledger error" is a String here:

https://github.com/mobilecoinfoundation/mobilecoin/blob/f142750ae329fd94bb0cbc480d4d8019bf182763/transaction/core/src/validation/error.rs#L116

There is a lot of stuff that is crappy about that conversions module -- for instance, if we want to convert every error into an error code, maybe we can include the text in the body of the response? And we could try to convert an error code + response body back into this error I guess, so that the conversion is not lossy?

I suppose that instead of putting `String` here we could give `TxOutConversionError` and `AmountError` some default states which are "unknown" or something, but this also seems crappy.

If you have ideas I'm happy to try to improve it in another PR. This part of the diff was actually already merged in the masked_amount_v2 PR and I haven't yet rebased this PR on master. But the review is helpful and I'm going to make a PR to address the follow on issues.

Reply:
I see, yes, that is annoying. Thanks for explaining it.
I made an attempt [here](https://github.com/mobilecoinfoundation/mobilecoin/blob/master/consensus/api/proto/consensus_client.proto#L32) to try and support a bit less lossy error conversions (also see [this](https://github.com/mobilecoinfoundation/mobilecoin/blob/master/consensus/api/src/conversions.rs#L206) for example of how that stuff got used). Not sure that is ideal either... I think losing the error details is going to be the thing that we get upset with the moment we run into these errors and are left wondering what happened, but this is a problem that is outside the scope of this PR.

Reply:
It looks that we only use this conversion in `mc-connection` here:

https://github.com/mobilecoinfoundation/mobilecoin/blob/bae5a20a59e54c5a288a1772971684b3b37a351a/connection/src/error.rs#L92

And we are only marshalling back to `TransactionValidationError` to satisfy this:

https://github.com/mobilecoinfoundation/mobilecoin/blob/bae5a20a59e54c5a288a1772971684b3b37a351a/connection/src/error.rs#L35

I wonder if we can just change that to `ProposeTxResult` and be done with it?

Reply:
I think  the reason the conversion from our custom error codes to a Rust `Result::Err` variant take place is in order to essentially obfuscate the fact that we are doing an RPC call, and make the caller experience be similar to had they called propose_tx directly. It's possible this can all be simplified if https://github.com/mobilecoinfoundation/mobilecoin/blob/bae5a20a59e54c5a288a1772971684b3b37a351a/connection/src/thick.rs#L403 is changed to return the full RPC result (https://github.com/mobilecoinfoundation/mobilecoin/blob/bae5a20a59e54c5a288a1772971684b3b37a351a/consensus/api/proto/consensus_common.proto#L99) and have callers handle it appropriately. I think the original thinking was that the callers would have a better experience if they got a proper `Result::Err` when the transaction was declined (but the GRPC call succeeded).

Reply:
So like, what I was thinking was, if that enum option was just:

```
 TransactionValidation(TxProposeResult), 
```

then we don't need to do the conversion at all and that saves us a headache? And the caller still has a "normal" rust error I think, so it shouldn't be that disruptive.

I can understand that like, maybe the error strings for `ProposeTxResponse` are worse than for `TransactionValidationError` if you log it? But maybe we can simply improve them if that's the issue?

Reply:
I opened a PR showing how I think this should work and why: https://github.com/mobilecoinfoundation/mobilecoin/pull/2501